m2cgen/interpreters/java/interpreter.py

```diff
@@ -28,58 +30,52 @@ def interpret(self, expr):
             top_cg.add_package_name(self.package_name)
 
         with top_cg.class_definition(self.model_name):
-            if isinstance(expr, ast.SubroutineExpr):
-                self._subroutine_expr_queue = []
-                self._do_interpret(expr)
-            else:
-                self._subroutine_expr_queue = [
-                    Subroutine("score", expr)
-                ]
-
-            while len(self._subroutine_expr_queue) > 0:
-                self._process_next_subroutine()
-                top_cg.add_code_lines(self._cg.code)
+            while len(self._subroutine_expr_queue):
+                subroutine_code = self._process_next_subroutine()
+                top_cg.add_code_lines(subroutine_code)
 
         return [
             (self.model_name, top_cg.code),
         ]
 
-    def interpret_subroutine_expr(self, expr, **kwargs):
-        method_name = self._get_subroutine_name()
-        return self._enqueue_subroutine(method_name, expr)
-
-    def interpret_main_expr(self, expr, **kwargs):
-        return self._enqueue_subroutine("score", expr)
-
     def interpret_array_expr(self, expr, **kwargs):
         nested = []
         for e in expr.exprs:
             nested.append(self._do_interpret(e, **kwargs))
         return self._cg.array_init(nested)
 
+    def _create_code_generator(self):
+        return JavaCodeGenerator(indent=self.indent)
+
+    # Methods to support ast.SubroutineExpr
+
+    def interpret_subroutine_expr(self, expr, **kwargs):
+        method_name = self._get_subroutine_name()
+        return self._enqueue_subroutine(method_name, expr)
+
     def _enqueue_subroutine(self, name, expr):
         self._subroutine_expr_queue.append(Subroutine(name, expr.expr))
         return self._cg.method_invocation(name, self._feature_array_name)
 
     def _process_next_subroutine(self):
         subroutine = self._subroutine_expr_queue.pop(0)
-        is_multi_output = subroutine.expr.is_multi_output
-        return_type = "double[]" if is_multi_output else "double"
+        _is_vector_output = subroutine.expr._is_vector_output
+
         self._cg = self._create_code_generator()
 
         with self._cg.method_definition(
                 name=subroutine.name,
-                args=[("double[]", self._feature_array_name)],
-                return_type=return_type):
+                args=[
+                    (True, self._feature_array_name)],
```

Comment:
I'm not sure I like the way `args` parameter looks now. Before it represented type -> value. I'm not sure it generalizes well. Are we ever going to generate methods with arguments other than `double` and `double[]`?